cmd/ip.go

```diff
@@ -73,89 +68,86 @@ func (s ipManager) getAllIPs() []net.IP {
 	return ips
 }
 
-func (s ipv6Stack) filterIPs(ips []net.IP) []net.IP {
-	// ip.IsGlobalUnicast() returns true for:
-	// GUA = Global Unicast Address
-	// ULA = Unique Local Address
-	// We prefer GUA over ULA.
-	ipv6s := []net.IP{}
-	for _, ip := range ips {
-		if ip.IsGlobalUnicast() && ip.To4() == nil {
-			ipv6s = append(ipv6s, ip)
-		}
-	}
-
-	return ipv6s
-}
+// The score function evaluates the value of a given IP address and returns
+// a score of type uint64.
+// The higher the score, the more valuable the IP address.
+func (s ipManager) score(ip net.IP) uint64 {
+	// Score format:
+	// +------------+------------------+--------------+
+	// | Reserved   | Priority Subnets | Address Type |
+	// | (28 bits)  |    (32 bits)     |   (4 bits)   |
+	// +------------+------------------+--------------+
+	score := uint64(0)
 
-func (s ipv4Stack) filterIPs(ips []net.IP) []net.IP {
-	ipv4s := []net.IP{}
-	for _, ip := range ips {
+	// Scoring by IP address type
+	if len(ip) == 16 {
 		if ip.To4() != nil {
-			ipv4s = append(ipv4s, ip)
+			return 0
 		}
+		if !ip.IsGlobalUnicast() {
+			return 0
+		}
+		if ipv6IsGUA(ip) {
+			score += 0x8
+			if ipv6IsEUI64(ip) {
+				score += 0x4
+			}
+		}
+	} else if len(ip) == 4 {
+		if ip.To4() == nil {
+			return 0
+		}
+		if ip.IsGlobalUnicast() {
+			score += 0x8
+		}
+		if ipv4IsSAS(ip) {
+			score += 0x4
+		}
+		if ip.IsPrivate() {
+			score += 0x2
+		}
+	} else {
+		return 0
 	}
+	score += 0x1
```

Comment:
This line is a little confusing. Let's initialize `score` with `0x1` from the start.